File: 2026.07/2026.07.03/3304-find-the-k-th-character-in-string-game-i.py
class Solution:
    def kthCharacter(self, k: int) -> str:
        # Solution: Binary Search + Count
        # Time: O(logN)
        # Space: O(1)

        n = 256
        count = 0

        while k > 1:
          if n < k:
              k -= n
              count += 1
          n = n / 2

        count = count % 26

        return chr(ord('a') + count)

print(Solution().kthCharacter(1))  # a
print(Solution().kthCharacter(2))  # b
print(Solution().kthCharacter(3))  # b
print(Solution().kthCharacter(4))  # c
print(Solution().kthCharacter(5))  # b
print(Solution().kthCharacter(6))  # c
print(Solution().kthCharacter(7))  # c
print(Solution().kthCharacter(8))  # d

File: 2026.07/2026.07.03/3304-find-the-k-th-character-in-string-game-i.js
/**
 * @param {number} k
 * @return {character}
 */
var kthCharacter = function (k) {
  // Solution: Binary Search + Count
  // Time: O(logN)
  // Space: O(1)

  let n = 256;
  let count = 0;

  while (k > 1) {
    if (n < k) {
      k -= n;
      count++;
    }

    n = n / 2;
  }

  count = count % 26;
  return String.fromCharCode(97 + count);
};

// a
// ab
// abbc
//  ^ +1
// abbcbccd
//      ^

// index 4 => index 2 + 1 => index 1 + 2 => c
// index 5 => index 1 + 1 => b
// index 6 => index 2 + 1 => index 1 + 2 => c

console.log(kthCharacter(1)); // a
console.log(kthCharacter(2)); // b
console.log(kthCharacter(3)); // b
console.log(kthCharacter(4)); // c
console.log(kthCharacter(5)); // b
console.log(kthCharacter(6)); // c
console.log(kthCharacter(7)); // c
console.log(kthCharacter(8)); // d
